### ISSUE TITLE
GPU memory leakage when using python binding to generate AOT-compiled GPU code

### ISSUE THREAD

**Issue by yyuting** [halide/Halide#5883]
I’m having GPU memory leakage when using the python binding to generate AOT-compiled GPU code. The memory usage keeps increasing after each call to the compiled function, and eventually raising an out-of-memory error. I slightly modified lesson 10 of the original tutorial to generate AOT-compiled GPU code as in the files below, and reached OOM error after executing the compiled function 90 times.

I'm using Linux + Nvidia GPU, and the target is x86-64-linux-avx-avx2-cuda-f16c-fma-sse41. I tried both Halide 10.0.0 and the current master branch, both have the memory leakage problem. But the C++ interface never had these problems. Is there a bug in the python binding or am I doing something wrong when generating GPU code in python? Thanks!

lesson_10_aot_compilation_generate_gpu.py

    import halide as hl

    def main():

        brighter = hl.Func("brighter")
        x, y = hl.Var("x"), hl.Var("y")

        # Define the hl.Func.
        brighter[x, y] = hl.cast(hl.Float(32), x * y)

        uo, ui, vo, vi = hl.Var('uo'), hl.Var('ui'), hl.Var('vo'), hl.Var('vi')
        brighter.compute_root().gpu_tile(x, y, uo, vo, ui, vi, 32, 8)

        target = hl.get_host_target().with_feature(hl.TargetFeature.CUDA)

        fname = "lesson_10_halide_gpu"
        brighter.compile_to({hl.Output.object: "lesson_10_halide_gpu.o",
                             hl.Output.python_extension: "lesson_10_halide_gpu.py.cpp"},
                            [], "lesson_10_halide_gpu",
                            target=target)

        print("Halide pipeline compiled, but not yet run.")

        return 0

    if __name__ == "__main__":
        main()

lesson_10_aot_compilation_run_gpu.py

    import lesson_10_halide_gpu

    import numpy as np
    import halide as hl

    def main():

        output = np.empty((6400, 4800), dtype=np.float32, order='F')
        output_img = hl.Buffer(output)

        output_img.copy_to_host()
        for i in range(1000):
            lesson_10_halide_gpu.lesson_10_halide_gpu(output)
            output_img.device_sync()
            print('running %d iterations' % i)
        output_img.copy_to_host()

        # Everything worked!
        print("Success!")
        return 0

    if __name__ == "__main__":
        main()

**Comment by ryanstout** [comment]
@yyuting were you able to figure anything out on this issue? I'm seeing the same thing.

**Comment by ryanstout** [comment]
seeing the same thing on OpenCL on my end

**Comment by yyuting** [comment]
no, I ended up writing and compiling the kernel in C++, then using pybind11 to return the output buffer to python.

**Comment by ryanstout** [comment]
@yyuting I spent today on this, too late I'm guessing, but I made some changes to the python bindings and have it not leaking now. I'll do a PR this week.

**Comment by yyuting** [comment]
thx!

**Comment by abadams** [comment]
> [!NOTE]
> **Automated triage:** this issue was flagged by Claude as likely safe to close, for the reason below. The reasoning was checked against the current source, and a second pass tried and failed to refute it, but no human verified this specific text. **If the reason is wrong, please reopen the issue and say so** — some fraction of these are expected to be wrong.

**Reason (CLOSE_FIXED):** The leak was in the generated Python extension wrapper: each call wrapped the numpy array in a temporary `halide_buffer_t` whose device allocation was never released. `PyHalideBuffer`'s destructor now calls `halide_device_free()`.

**Evidence:** Commit `1653c16b0` "PythonExtensionGen: ~PyHalideBuffer should call device_free() (#8399) (#8439)" (2024-10-21); current code at `src/PythonExtensionGen.cpp:359-362`.

Fixed by commit `1653c16b0` ("PythonExtensionGen: ~PyHalideBuffer should call device_free()", #8439): the generated extension's `PyHalideBuffer` destructor now calls `halide_device_free()` on buffers it allocated device memory for (`src/PythonExtensionGen.cpp:359-362`). Closing as fixed.
